### DISCUSSION TITLE
Script to show DB schema in human-readable form

### DISCUSSION THREAD

**Discussion by Lex-2008** [pocketbase/pocketbase#252] [Show and tell]
While "proper" import/export of database schema is in works (#48), we can use [collections API](https://pocketbase.io/docs/api-collections/) to get a JSON describing all collections and all fields. I made a script to make it more suitable for human consumption - useful when "import" of schema is implemented by manually clicking in the admin UI. Also it tries to be diff-friendly: by seeing a `diff` of two outputs of this script, a human is likely to understand what needs to be changed in order to make one DB schema match the other.

[humanschema.js](https://github.com/pocketbase/pocketbase/files/9235134/humanschema.js.txt) is attached to this discussion, and used like this:
```html
<!doctype html><html><head><title>Show schema</title>
</head><body>
<pre id="out">loading...</pre>
<!-- <script src="../lib/pocketbase.umd.min.js"></script> -->
<script src="https://cdn.jsdelivr.net/npm/pocketbase/dist/pocketbase.umd.min.js"></script>
<script src="humanschema.js"></script>
<script>
// const client = new PocketBase('/');
const client = new PocketBase('https://pocketbase.io/');

client.Collections.getFullList(200 /* batch size */, { sort: "name"}).then(
        collections=>document.getElementById('out').innerHTML=collections.map(prettyCollection).join('\n\n'),
        error=>console.error(error));

</script></body></html>
```

For the [pocketbase demo](https://pocketbase.io/demo/) database, it produces this output:
```
collection messages:
  listRule: @request.user.id != null
  viewRule: @request.user.id != null
  createRule: @request.user.id != null
  updateRule: author = @request.user.id
  deleteRule: author = @request.user.id
  field message: type text
    required
  field author: type user
    Delete record on relation delete: true
    required

collection posts:
  listRule: active = true
  viewRule: active = true
  createRule: admins only (green padlock)
  updateRule: admins only (green padlock)
  deleteRule: admins only (green padlock)
  field title: type text
    max: 100
    required
  field description: type text
  field active: type bool
  field options: type select
    choices: optionA,optionB,optionC
    maxSelect: 2
  field featuredImages: type file
    maxSize: 5242880
    maxSelect: 3
    mimeTypes: NO_UPLOADS_ALLOWED
    thumbs: 100x200

collection profiles:
  listRule: userId = @request.user.id
  viewRule: userId = @request.user.id
  createRule: userId = @request.user.id
  updateRule: userId = @request.user.id
  deleteRule: admins only (green padlock)
  field userId: type user
    Delete record on relation delete: true
    required
    unique
  field name: type text
  field avatar: type file
    maxSize: 5242880
    mimeTypes: NO_UPLOADS_ALLOWED
  field website: type url
```

If someone else finds this useful - feel free to use it and modify however you want!

**Comment by benallfree** [comment]
Added to https://github.com/benallfree/awesome-pocketbase :)

**Reply by ganigeorgiev** [reply]
@benallfree _Just for clarity, there is already JSON export/import for collections configuration from the Admin UI, including automated diff ("Settings > Export/Import collections")._

**Reply by benallfree** [reply]
Thank you! Yes as it happens I was playing with that export a few days ago because I was exploring generating Typescript definitions from it. Very helpful!
